src/test/java/com/emc/object/s3/S3JerseyClientTest.java

```diff
@@ -500,6 +500,89 @@ public void testDeleteBucketWithObjects() throws Exception {
+        } catch (S3Exception e) {
+            Assert.assertEquals(403, e.getHttpCode());
+            Assert.assertEquals("EmptyBucketInProgress", e.getErrorCode());
+        }
+        try {
+            client.putObject(bucketName, "bar", "foo", null);
+            Assert.fail("It is not allowed to update a bucket whose background deletion tasks are in progress.");
+        } catch (S3Exception e) {
+            Assert.assertEquals(403, e.getHttpCode());
+            Assert.assertEquals("EmptyBucketInProgress", e.getErrorCode());
+        }
+
+        // Empty bucket execution frequency is 1 min, so 2 min would be enough.
+        // Wait until deletion background tasks are DONE.
+        Thread.sleep(2 * 60 * 1000);
+
+        try {
+            client.getBucketDeletionStatus(bucketName);
+            Assert.fail("GET EmptyBucketStatus when the bucket is deleted should give 404.");
+        } catch (S3Exception e) {
+            Assert.assertEquals(404, e.getHttpCode());
+            Assert.assertEquals("The specified bucket does not exist", e.getMessage());
+        }
+
+        Assert.assertFalse("Failed to delete bucket" + bucketName, client.bucketExists(bucketName));
+    }
+
+    @Test
+    public void testDeleteBucketInRetentionWithBackgroundTasks() throws Exception {
+        Assume.assumeTrue("ECS version must be at least 3.8", ecsVersion != null && ecsVersion.compareTo("3.8") >= 0);
+        Assume.assumeTrue("Skip Object Lock related tests for non IAM user.", isIamUser);
+
+        String bucketName = getTestBucket() + "-InRetention";
+        client.createBucket(bucketName);
+        String key = "testObject_PutObjectRetention";
+        client.enableObjectLock(bucketName);
+        // ensure the retention period is longer than the bucket deletion background tasks.
+        Date retentionDate = new Date(System.currentTimeMillis() + 150000);
+        ObjectLockRetention objectLockRetention = new ObjectLockRetention()
+                .withMode(ObjectLockRetentionMode.COMPLIANCE)
+                .withRetainUntilDate(retentionDate);
+        PutObjectRequest putObjectRequest = new PutObjectRequest(bucketName, key, "test Put Object Retention")
+                .withObjectMetadata(new S3ObjectMetadata().withObjectLockRetention(objectLockRetention));
+        client.putObject(putObjectRequest);
+
+        client.deleteBucket(new DeleteBucketRequest(bucketName, true));
+        // Empty bucket execution frequency is 1 min, so 2 min would be enough.
+        // Wait until deletion background tasks are DONE.
+        Thread.sleep(2 * 60 * 1000);
+
+        BucketDeletionStatus status = client.getBucketDeletionStatus(bucketName);
+        // check the status and dates of the bucket delete task, other entries are not given.
+        Assert.assertEquals("The EmptyBucketStatus status of a bucket which is still in retention should be FAILED.", status.getStatus(), "FAILED");
```

Comment:
create multiple objects (i.e. include one without retention) and check other error fields to make sure they are populated (i.e. FailedToDeleteDueToRetention and EntriesDeleted)